src/saltext/vmware/modules/vmc_vm_disks.py

```diff
@@ -0,0 +1,480 @@
+        error_msg = "Mandatory params {} are missing from user input".format(validation_result)
+        log.error(error_msg)
+        return {vmc_constants.ERROR: error_msg}
+
+    api_url_base = vmc_request.set_base_url(hostname)
+    api_url = "{base_url}api/vcenter/vm/{vm_id}/hardware/disk/{disk_id}"
+    api_url = api_url.format(base_url=api_url_base, vm_id=vm_id, disk_id=disk_id)
+
+    headers = vmc_vcenter_request.get_headers(hostname, username, password)
+
+    payload = {"backing": {"type": backing_type, "vmdk_file": vmdk_file}}
+
+    return vmc_vcenter_request.call_api(
+        method=vmc_constants.PATCH_REQUEST_METHOD,
+        url=api_url,
+        headers=headers,
+        description="vmc_vm_disks.update",
+        responsebody_applicable=False,
+        verify_ssl=verify_ssl,
+        cert=cert,
+        data=payload,
+    )
+
+
+def create(
+    hostname,
+    username,
+    password,
+    vm_id,
+    bus_adapter_type,
+    verify_ssl=True,
+    cert=None,
+    backing=None,
+    new_vmdk=None,
+    scsi=None,
+    sata=None,
+    ide=None,
+):
+    """
+    Creates virtual disk for given VM.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt vm_minion vmc_vm_disks.create hostname=sample-vcenter.vmwarevmc.com ...
+
+    hostname
+       Hostname of the vCenter console
+
+    username
+        username required to login to vCenter console
+
+    password
+        password required to login to vCenter console
+
+    vm_id
+        Virtual machine identifier for which the disk should be added.
+
+    bus_adapter_type
+        Type of host bus adapter to which the disk should be attached. Possible values are: IDE, SATA, SCSI
+
+    verify_ssl
+        (optional) Option to enable/disable SSL verification. Enabled by default.
+        If set to False, the certificate validation is skipped.
+
+    cert
+        (optional) Path to the SSL client certificate file to connect to VMC Cloud Console.
+        The certificate can be retrieved from browser.
+
+    backing
+        (optional) Existing physical resource backing for the virtual disk.
+        Exactly one of backing or new_vmdk must be specified.
+        If unset, the virtual disk will not be connected to an existing backing.
+
+        It is a json object which can contain 'type' and 'vmdk_file' keys.
+
+        'type': (String) (mandatory)
+            Backing type for the virtual disk. Possible values are: VMDK_FILE
+
+        'vmdk_file': (String) (optional)
+            Path of the VMDK file backing the virtual disk.
+            This field is optional and it is only relevant when the value of backing type is VMDK_FILE
+
+        For ex:
+
+            .. code::
+
+                "backing": {
+                    "type": "VMDK_FILE",
+                    "vmdk_file": "[WorkloadDatastore] 8dff8760-b6da-24d2-475b-0200a629f7fc/test-vm-1.vmdk"
+                }
+
+    new_vmdk
+        (optional) Specification for creating a new VMDK backing for the virtual disk.
```

Comment:
**question** Would a typical VMC user expect to use backing vs vmdk? This feels awkward to me. I would just expect to provide the criteria that are required. In fact, since there's no other type available for backing why not just have the API just look something like this?

```
create
   (Optional) whether or not to create the vmdk_file if it does not exist. Defaults to True.

vmdk_file
    (Optional) A base name for the physical resource backing for the disk, or an existing vmdk file. If the vmdk_file does not exist, it will be created, unless ``create`` is False. If ``vmdk_file`` is not provided, the name will be chosen by the server, derived from the name of the virtual machine.
     
capacity
    (Optional) The capacity for the disk, in bytes. Ignored if the vmdk file already exists.

storage_policy_id
    (Optional) The identifier for the storage policy to be used when creating the vmdk file.
```

Maybe there's a good reason to do it this way, but if we can simplify the options we should.

Reply:
> **suggestion** same comment about optional. This is required unless `backing` is provided.

updated

Reply:
> **question** Would a typical VMC user expect to use backing vs vmdk? This feels awkward to me. I would just expect to provide the criteria that are required. In fact, since there's no other type available for backing why not just have the API just look something like this?
> 
> ```
> create
>    (Optional) whether or not to create the vmdk_file if it does not exist. Defaults to True.
> 
> vmdk_file
>     (Optional) A base name for the physical resource backing for the disk, or an existing vmdk file. If the vmdk_file does not exist, it will be created, unless ``create`` is False. If ``vmdk_file`` is not provided, the name will be chosen by the server, derived from the name of the virtual machine.
>      
> capacity
>     (Optional) The capacity for the disk, in bytes. Ignored if the vmdk file already exists.
> 
> storage_policy_id
>     (Optional) The identifier for the storage policy to be used when creating the vmdk file.
> ```
> 
> Maybe there's a good reason to do it this way, but if we can simplify the options we should.

Yes the user is expected to use either backing or new_vmdk.  if he wants to use any existing vmdk file, he can pass value to the `backing` param ( that contains the type and path to that physical backing resource).
if he wants to use new physical resource, he can pass value to the `new_vmdk` (with the required specifications. we can also pass empty json, then vcenter will create the physical backing resource with default specifications ).

Reply:
Since these are mutually exclusive, and unambiguous we should make this change - I'd be happy to pair on Wednesday morning my time/evening your time.

Reply:
sure @waynew, i will schedule call for wednesday 7:30 pm IST